### PR TITLE
Re-organise the kernel tree in sidebar

### PR DESCRIPTION
**Pull request by krassowski**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/15294

## Code changes

- Adds `KernelSpecItem` which wraps `RunningKernel` items.
- `RunningKernel` no longer includes kernelspecs
- CSS is cleaned up
- [x] resize down the collapse/expand buttons to match visually with list/tree toggle
- [x] add a setting to remember preference list/tree
- [x] fix behaviour of buttons overflow when sidebar is resized down
- [x] update snapshots in tests

## User-facing changes

Early draft - to allow for more  interactive feedback

collapse-expand-list-view

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by krassowski** [comment]
Now that there are more buttons a visual glitch appeared:
- "Kernels" sometimes get ellipsed to "Kern..." when the sidebar is resized down
- more importantly the "Shut Down All" buttons gets hidden because it moves to new line in overflow

I experimented with hiding the expand/collapse buttons which is now possible using CSS only with container queries:

```css
.jp-RunningSessions.jp-SidePanel {
  /* Allow defining container queries when embedded in sidebar (which already uses strict size containment) */
  container-type: size;
  container-name: jp-running-sessions-side-panel;
}

@container jp-running-sessions-side-panel (max-width: 250px) {
  .jp-RunningSessions-collapseButton {
    /* hide if sidebar too narrow */
    display: none;
  }
}
```


But given that we only need to hide it on very small distance for the default usage with "Kernels" title I feel like this is an overkill and might be more confusing than helpful ("why is it hiding when I resize it down - there _seems_ to be enough space").

Instead I am shrinking the margins of these two collapse/expand buttons slightly which will bring them together (which makes sense to me as it will visually grouping them); and to prevent the "Shut Down All" button from dropping I am setting `min-width: max-content;` on the toolbar:

| Problem | Hiding collapse/expand buttons | Final Solution |
|--|--|--|
| the-problem | experiment | min-width |

**Comment by krassowski** [comment]
This is ready for review. I am not convinced whether there is enough benefit to warrant the expand/collapse buttons but I will let @ericsnekbytes who suggested these let experiment with the functional version and provide feedback.

**Comment by ericsnekbytes** [comment]
Conceptually, I think expand/collapse are good ideas for these kind of multi-notebook/console/shared-kernel workflows. Some things to note during my testing:

- Expand and collapse all are only valuable when the expand/collapse state are remembered per item in the tree. Right now, it seems expanding one item also expands all its children, which defeats the purpose of collapsing all to reduce noise and to allow users to individually drill down only into items that they care to expand/manage at that point.
- Remembering which sections are expanded/collapsed would be an essential part of navigating and managing open kernels/notebooks/sessions in this UI and with these multi-kernel/shared-kernel workflows...
- Right clicking the notebook tab/name and starting "new console for notebook" does not show the new console in the UI here (it seems to only work when right clicking the kernel entry and making a new console from this UI)

The UI and experience look good, and successfully communicate which kernels share state 👍  Expand/collapse all will not provide much value without individual and persistent expand/collapse state for individual tree entries.

Thanks for working on this!

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @krassowski 

A couple of comments/questions:

- When opening a console linked to a notebook from the context menu on the notebook tab, the console is not listed. Is it expected?

image

- For expanding/collapsing all, in the toc with use a unique button with different icons than the one used here. I propose to be homogeneous by using one or the other (no preference)

- Similarly for switching between list and tree, in the debugger for switching between table and tree, the two buttons are always displayed and the selected mode has emphasis. Being homogeneous would be better (again no preference for one or the other approach)

- Should we display the kernel id differently - like using lighter color ?

**Comment by krassowski** [comment]
Thanks @fcollonval

> When opening a console linked to a notebook from the context menu on the notebook tab, the console is not listed. Is it expected?

No, this is not expected, good catch. It looks this is pre-existing issue unrelated to this PR - I opened https://github.com/jupyterlab/jupyterlab/issues/15960 to track it. My guess is that consoles created using `notebook:create-console` are not getting added to the tracker.

> For expanding/collapsing all, in the toc with use a unique button with different icons than the one used here

I think the icons from the ToC might be better here indeed. I switched icons in https://github.com/jupyterlab/jupyterlab/pull/15845/commits/f4d259725555db876aad079c82f0ed299d1964bf and made it into a unique button in https://github.com/jupyterlab/jupyterlab/pull/15845/commits/8a65c32bf6788f90c1311d7518431de195e8e124 

> Similarly for switching between list and tree, in the debugger for switching between table and tree, the two buttons are always displayed and the selected mode has emphasis

I would prefer to have a single icon here. If it makes sense to propagate this to the debugger is not clear to me but if homogeneity is the guiding principle then I would go that way. Happy to hear counterpoints though.

> Should we display the kernel id differently - like using lighter color ?

Done in https://github.com/jupyterlab/jupyterlab/pull/15845/commits/f4d259725555db876aad079c82f0ed299d1964bf

File: packages/running-extension/src/kernels.tsx [hunk 1]
```diff
@@ -224,20 +290,26 @@ namespace Private {
...
+              '%1 and %2 more',
+              children[0].label(),
+              children.length - 1
+            );
+      const kernelIdPrefix = kernel.id.split('-')[0];
+      return (
+        <>
+          {summary} <span className={KERNEL_LABEL_ID}>({kernelIdPrefix})</span>
+        </>
+      );
```
Comment (fcollonval) [inline packages/running-extension/src/kernels.tsx:312]: This breaks `labelTitle` (see line 318 ```const title = [`${this.label()}: ${id}`];```):

image

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @krassowski 

I spotted a minor bug and I have a question about the table view; should we display the kernel id in a second column (a bit like last modified in the file browser)? Otherwise the kernels section does not bring lots of information compare to the open tabs.
> That can be a follow-up

File: packages/running-extension/src/kernels.tsx (hunk 1 above)
Comment (krassowski) [reply packages/running-extension/src/kernels.tsx]: Thanks, fixed in 5f10b30.

**Comment by krassowski** [comment]
> should we display the kernel id in a second column (a bit like last modified in the file browser)?

I do like this idea. I opened https://github.com/jupyterlab/jupyterlab/issues/15991 to track it.

**Review by fcollonval — APPROVED** [review]
Thanks @krassowski

**Pull request merged**